scio-core/src/main/scala/com/spotify/scio/sql/Sql.scala

```diff
@@ -0,0 +1,341 @@
+import org.apache.beam.sdk.extensions.sql.impl.rel.BeamRelNode
+import org.apache.beam.sdk.extensions.sql.impl.{BeamSqlEnv, ParseException}
+import org.apache.beam.sdk.extensions.sql.impl.schema.BaseBeamTable
+import org.apache.beam.sdk.extensions.sql.impl.utils.CalciteUtils
+
+import scala.language.experimental.macros
+import scala.util.Try
+import scala.collection.JavaConverters._
+
+final case class Query[A, B](query: String, udfs: List[Udf] = Nil)
+
+final case class Query2[A, B, R](query: String,
+                                 aTag: TupleTag[A],
+                                 bTag: TupleTag[B],
+                                 udfs: List[Udf] = Nil)
+
+object Queries {
+
+  /**
+   * Typecheck [[Query]] q against the provided schemas.
+   * If the query correctly typechecks, it's simply return as a [[Right]].
+   * If it fails, a error message is returned in a [[Left]].
+   */
+  def typecheck[A: Schema, B: Schema](q: Query[A, B]): Either[String, Query[A, B]] = {
+    val schema: BSchema = SchemaMaterializer.fieldType(Schema[A]).getRowSchema
+    val expectedSchema: BSchema =
+      Schema[B] match {
+        case s: Record[B] =>
+          SchemaMaterializer.fieldType(s).getRowSchema
+        case _ =>
+          SchemaMaterializer.fieldType(Schema[ScalarWrapper[B]]).getRowSchema
+      }
+
+    QueryUtils.typecheck(q.query, schema, expectedSchema).right.map(_ => q)
+  }
+
+  /**
+   * Typecheck [[Query2]] q against the provided schemas.
+   * If the query correctly typechecks, it's simply return as a [[Right]].
+   * If it fails, a error message is returned in a [[Left]].
+   */
+  def typecheck[A: Schema, B: Schema, R: Schema](
+    q: Query2[A, B, R]): Either[String, Query2[A, B, R]] = {
+    val schemaA: BSchema = SchemaMaterializer.fieldType(Schema[A]).getRowSchema
+    val schemaB: BSchema = SchemaMaterializer.fieldType(Schema[B]).getRowSchema
+    val expectedSchema: BSchema =
+      Schema[R] match {
+        case s: Record[R] =>
+          SchemaMaterializer.fieldType(s).getRowSchema
+        case _ =>
+          SchemaMaterializer.fieldType(Schema[ScalarWrapper[R]]).getRowSchema
+      }
+
+    QueryUtils
+      .typecheck(q.query, List((q.aTag.getId, schemaA), (q.bTag.getId, schemaB)), expectedSchema)
+      .right
+      .map(_ => q)
+  }
+
+  def typedQuery[A: Schema, B: Schema](query: String, udfs: Udf*): Query[A, B] =
+    macro QueryMacros.typedQuerylImpl[A, B]
+
+  def typedQuery[A: Schema, B: Schema, R: Schema](query: String,
+                                                  aTag: TupleTag[A],
+                                                  bTag: TupleTag[B],
+                                                  udfs: Udf*): Query2[A, B, R] =
+    macro QueryMacros.typedQuery2Impl[A, B, R]
+}
+
+object Sql {
+
+  def from[A: Schema](sc: SCollection[A]): SqlSCollection[A] = new SqlSCollection(sc)
+
+  def from[A: Schema, B: Schema](a: SCollection[A], b: SCollection[B]): SqlSCollection2[A, B] =
```

Comment:
Maybe the tag should be there ? Something like this would make sense to me:

```scala
Sql.from("a" -> sc1, "b" -> sc2)
  .query(Query("..."))
```

wdyt ?

Reply:
Ideally we'd keep track of the required binding in the type, but I'm not sure that's possible:

```scala
val q: Query[('a, User), ('b, Playlist), Int]("SELECT foo FROM a INNER JOIN b on ...)
```

Reply:
yeah that would be nice, but it's not really possible to have `Symbol`'s like that :( also literal types (using shapeless) can be too much?

Reply:
Definitelly too much. I wish there was a way to statically check that the name you give to the `SCollection` matches the SQL query but it's probably impossible.